qubes/backup.py

```diff
@@ -630,39 +630,44 @@ async def _wrap_and_send_files(self, files_to_backup, output_queue):
                 path = file_info.path
                 if callable(path):
                     path = await qubes.utils.coro_maybe(path())
-                tar_cmdline = (["tar", "-Pc", '--sparse',
-                                '-C', os.path.dirname(path)] +
-                               (['--dereference'] if
-                               file_info.subdir != "dom0-home/" else []) +
-                               ['--xform=s:^%s:%s\\0:' % (
-                                   os.path.basename(path),
-                                   file_info.subdir),
-                                   os.path.basename(path)
-                               ])
-                file_stat = os.stat(path)
-                if stat.S_ISBLK(file_stat.st_mode) or \
-                        file_info.name != os.path.basename(path):
-                    # tar doesn't handle content of block device, use our
-                    # writer
-                    # also use our tar writer when renaming file
-                    assert not stat.S_ISDIR(file_stat.st_mode), \
-                        "Renaming directories not supported"
-                    tar_cmdline = ['python3', '-m', 'qubes.tarwriter',
-                        '--override-name=%s' % (
-                            os.path.join(file_info.subdir, os.path.basename(
-                                file_info.name))),
-                        path]
-                if self.compressed:
-                    tar_cmdline.insert(-2,
-                        "--use-compress-program=%s" % self.compression_filter)
-
-                self.log.debug(" ".join(tar_cmdline))
-
-                # Pipe: tar-sparse | scrypt | tar | backup_target
-                # TODO: log handle stderr
-                # pylint: disable=not-an-iterable
-                tar_sparse = await asyncio.create_subprocess_exec(
-                    *tar_cmdline, stdout=subprocess.PIPE)
+
+                fd = os.open(path, os.O_RDONLY|o.O_NOCTTY|os.O_CLOEXEC)
+                try:
+                    fd_path = '/proc/self/fd/' + str(fd)
+                    tar_cmdline = (["tar", "-Pc", '--sparse',
+                                    '-C', os.path.dirname(path)] +
+                                   (['--dereference'] if
+                                   file_info.subdir != "dom0-home/" else []) +
+                                   ['--xform=s:^%s:%s\\0:' % (
```

Comment:
If you change how the argument looks like, you need to adjust this option too.

Reply:
Indeed I do.  That reminds me that the current code has a bug: `\`, `&`, and `:` in the second argument must be escaped.

Reply:
>  That reminds me that the current code has a bug: \, &, and : in the second argument must be escaped.

None of them can possibly appear in `file_info.subdir` (it's either `dom0-home/` or `vmX/`). If you _really_ need to do anything about it (which I think you don't), better add an assert, instead of possibly non-obvious escaping.

Reply:
> > That reminds me that the current code has a bug: , &, and : in the second argument must be escaped.
> 
> None of them can possibly appear in `file_info.subdir` (it's either `dom0-home/` or `vmX/`). If you _really_ need to do anything about it (which I think you don't), better add an assert, instead of possibly non-obvious escaping.

I will add an assert, mostly because the current code is _really_ non-obvious to someone (me) who has not looked at it before.